src/Serilog.Extensions.Logging/Extensions/Logging/EventIdPropertyCache.cs

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Serilog.Extensions.Logging
+{
+    using Microsoft.Extensions.Logging;
+    using Serilog.Events;
+
+    internal sealed class EventIdPropertyCache
+    {
+        private readonly object _createLock = new();
+        private readonly int _maxCapacity;
+        private readonly Dictionary<EventKey, LogEventProperty> _propertyCache;
+
+        private int count;
+
+        public EventIdPropertyCache(int maxCapacity)
+        {
+            _maxCapacity = maxCapacity;
+            _propertyCache = new Dictionary<EventKey, LogEventProperty>(capacity: maxCapacity);
+        }
+
+        public LogEventProperty GetOrCreateProperty(in EventId eventId)
+        {
+            var eventKey = new EventKey(eventId);
+
+            if (_propertyCache.TryGetValue(eventKey, out var cachedProperty))
```

Comment:
`Dictionary<,>` reads are not safe in the presence of concurrent writes; either `ConcurrentDictionary<,>` or the [non-generic `Hashtable` type](https://learn.microsoft.com/en-us/dotnet/api/system.collections.hashtable?view=net-8.0#thread-safety) work here. `Hashtable`'s a bit esoteric but might have better perf/lower contention.

Reply:
Based on the assumption that new `EventKey-LogEventProperty` pairs are not created often (but are rather retrieved from the cache), the non-generic `Hashtable` may not work well here. This is because the key is a struct, but `Hashtable` accepts `objects`, so the key would be boxed.

`ConcurrentDictionary` seems suitable for caching purposes.